launchpad/src/launchpad.rs

```diff
@@ -1,9 +1,519 @@
+
+        for ticket_id in first_ticket_id..=last_ticket_id {
+            let ticket_status = self.ticket_status().get(ticket_id);
+            if !ticket_status.is_winning(current_generation) {
+                continue;
+            }
+
+            self.set_ticket_status(
+                ticket_id,
+                TicketStatus::Confirmed {
+                    generation: current_generation,
+                },
+            );
+            actual_confirmed_tickets += 1;
+
+            if actual_confirmed_tickets == nr_tickets_to_confirm {
+                break;
+            }
+        }
+
+        require!(
+            actual_confirmed_tickets == nr_tickets_to_confirm,
+            "Couldn't confirm all tickets"
+        );
+
+        Ok(OperationCompletionStatus::Completed)
+    }
+
+    #[endpoint(claimLaunchpadTokens)]
+    fn claim_launchpad_tokens(&self) -> SCResult<OperationCompletionStatus> {
+        let caller = self.blockchain().get_caller();
+        require!(
+            !self.blacklist().contains(&caller),
+            "You have been put into the blacklist and may not claim tokens"
+        );
+
+        let current_epoch = self.blockchain().get_block_epoch();
+        let claim_start_epoch = self.claim_period_start_epoch().get();
+        let claim_period_in_epochs = self.claim_period_in_epochs().get();
+        let claim_end_epoch = claim_start_epoch + claim_period_in_epochs;
+
+        require!(
+            current_epoch > claim_start_epoch,
+            "Claim period has not started yet"
+        );
+
+        if current_epoch > claim_end_epoch {
+            self.select_new_winners()?;
+            return Ok(OperationCompletionStatus::InterruptedBeforeOutOfGas);
+        }
+
+        let (first_ticket_id, last_ticket_id) = self.try_get_ticket_range(&caller)?;
+        let current_generation = self.current_generation().get();
+        let mut nr_redeemed_tickets = 0;
+
+        for ticket_id in first_ticket_id..=last_ticket_id {
+            let ticket_status = self.ticket_status().get(ticket_id);
+            if !ticket_status.is_confirmed(current_generation) {
+                continue;
+            }
+
+            self.set_ticket_status(ticket_id, TicketStatus::Redeemed);
+            nr_redeemed_tickets += 1;
+        }
+
+        require!(nr_redeemed_tickets > 0, "No tickets to redeem");
+
+        self.nr_redeemed_tickets()
+            .update(|redeemed| *redeemed += nr_redeemed_tickets);
+
+        let launchpad_token_id = self.launchpad_token_id().get();
+        let tokens_per_winning_ticket = self.launchpad_tokens_per_winning_ticket().get();
+        let amount_to_send = Self::BigUint::from(nr_redeemed_tickets) * tokens_per_winning_ticket;
+
+        self.send()
+            .direct(&caller, &launchpad_token_id, 0, &amount_to_send, &[]);
+
+        Ok(OperationCompletionStatus::Completed)
+    }
+
+    // views
+
+    #[view(getNumberOfWinningTicketsForAddress)]
```

Comment:
A user may try to select_new_winners and query this view over and over again. If others do not pay attention, he'll be able to maximize his winning tickets. It this ok?

Reply:
The user has no control over how tickets are shuffled, and also has no way of influencing the shuffle.

Also, the shuffle is deterministic no matter the duration between `select_winners` calls. The actual shuffling of the ticket_ids is only done once, subsequent "select new winners" simply move the "range" of winning tickets from the `shuffled_tickets` (and they also automatically trigger a new "confirmation" and "claim" period, so no user can just endlessly select new winners until they win.

At least that was the intention.